### ISSUE TITLE
too many positional arguments in arm64 kvm

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#939]
```
arch/arm64/kvm/hyp/entry.S:112:87: error: too many positional arguments
 alternative_insn nop, .inst (0xd500401f | ((0) << 16 | (4) << 5) | ((!!1) << 8)), 4, 1
                                                                                      ^
```

**Comment by ihalip** [comment]
LLVM ASM parser doesn't like the space before the opening paranthesis. I don't like it, but this workaround does the trick:
```diff
diff --git a/arch/arm64/include/asm/sysreg.h b/arch/arm64/include/asm/sysreg.h
index b91570ff9db1..6dda7aa650e1 100644
--- a/arch/arm64/include/asm/sysreg.h
+++ b/arch/arm64/include/asm/sysreg.h
@@ -49,9 +49,9 @@
 #ifndef CONFIG_BROKEN_GAS_INST
 
 #ifdef __ASSEMBLY__
-#define __emit_inst(x)                 .inst (x)
+#define __emit_inst(x)                 .inst(x)
 #else
-#define __emit_inst(x)                 ".inst " __stringify((x)) "\n\t"
+#define __emit_inst(x)                 ".inst" __stringify((x)) "\n\t"
 #endif
 
 #else  /* CONFIG_BROKEN_GAS_INST */
```

**Comment by jcai19** [comment]
Reduced reproducer:

$ cat bad.s
.macro alternative_insn insn1
.endm

 alternative_insn .inst (0x0)



$ llvm-mc -triple=aarch64 bad.s -o /dev/null
bad.s:4:25: error: too many positional arguments
 alternative_insn .inst (0x0)

$ aarch64-linux-gnu-gcc -c bad.s
$ echo $?
0

Interestingly, on x86, gcc also fails to assemble

llvm-mc -triple=i386 bad.s -o /dev/null
bad.s:4:25: error: too many positional arguments
 alternative_insn .inst (0x0)

$ gcc -c bad.s
bad.s: Assembler messages:
bad.s:4: Error: too many positional arguments

**Comment by jcai19** [comment]
Took a look at LLVM's code, and it turned out the parsing of macro argument ([AsmParser::parseMacroArgument](https://llvm.org/doxygen/AsmParser_8cpp_source.html#l02577)) [stopped](https://llvm.org/doxygen/AsmParser_8cpp_source.html#l02604) after it read any space after .inst directive.  Interestingly, on x86, gcc also does not support such syntax.

$ cat bad.s
.macro alternative_insn insn
\insn
.endm

alternative_insn .space 8

**Comment by jcai19** [comment]
Having the integrated assembler to fully support directive calls as macro arguments seems to be a non-trivial task as we would have to parse all the directives with (sometimes optional) arguments. Unless such cases prevail in Linux code base, the proposed [workaround](https://github.com/ClangBuiltLinux/linux/issues/939#issuecomment-601776123 ) may be a better choice instead of having such support in LLVM IMO.

**Comment by jcai19** [comment]
On second thoughts,  the fix in the integrated assembler may be simpler than I thought. Send a patch for review at https://reviews.llvm.org/D76962.

**Comment by MaskRay** [comment]
It is worth mentioning that you may need to use psmith instead of peter.smith due to https://reviews.llvm.org/D73542#1871004 .

**Comment by jcai19** [comment]
Thanks! Will use the new username.

**Comment by MaskRay** [comment]
> LLVM ASM parser has trouble parsing that expression. I don't like it, but this workaround does the trick:
> 
> ```diff
> diff --git a/arch/arm64/include/asm/sysreg.h b/arch/arm64/include/asm/sysreg.h
> index b91570ff9db1..6dda7aa650e1 100644
> --- a/arch/arm64/include/asm/sysreg.h
> +++ b/arch/arm64/include/asm/sysreg.h
> @@ -49,9 +49,9 @@
>  #ifndef CONFIG_BROKEN_GAS_INST
>  
>  #ifdef __ASSEMBLY__
> -#define __emit_inst(x)                 .inst (x)
> +#define __emit_inst(x)                 .inst(x)
>  #else
> -#define __emit_inst(x)                 ".inst " __stringify((x)) "\n\t"
> +#define __emit_inst(x)                 ".inst" __stringify((x)) "\n\t"
>  #endif
>  
>  #else  /* CONFIG_BROKEN_GAS_INST */
> ```

```
--- i/arch/arm64/include/asm/sysreg.h
+++ w/arch/arm64/include/asm/sysreg.h
@@ -51,3 +51,3 @@
 #ifdef __ASSEMBLY__
-#define __emit_inst(x)                 .inst (x)
+#define __emit_inst(x)                 .inst(x)
 #else
```
may be sufficient.

**Comment by nickdesaulniers** [comment]
Ok, with the additional information:
1. https://reviews.llvm.org/D76962#1948922
2. https://github.com/ClangBuiltLinux/linux/issues/939#issuecomment-605408583
and additional vote for @ihalip 's sugguested change, I'm more supportive of fixing this on the Linux side.  Thanks @jcai19 and @MaskRay for digging into this more; sometimes the issues are quite nuanced, but I think @jcai19 's comment (link 1) is enough for me to think that we shouldn't imitate inconsistent behavior, as it's probably unintentional.  It might be worthwhile @jcai19 to file a bug against binutils with some examples demonstrating the inconsistency, and ask for clarity.  We might be able to come to a mutual solution with them.

**Comment by jcai19** [comment]
Thanks for the clarification. I have reported a bug against gas at https://sourceware.org/bugzilla/show_bug.cgi?id=25750, but don't know who I should cc so please feel free to add the right people.

**Comment by nickdesaulniers** [comment]
Thanks for reporting. Nick Clifton [suggests](https://sourceware.org/bugzilla/show_bug.cgi?id=25750#c1) a `:vararg` suffix.
1. does that solve the issue here when using `-no-integrated-as`?
2. does IA recognize `:vararg` suffix?

**Comment by MaskRay** [comment]
`:vararg` does not work for more than 1 argument. As I mentioned in the review, properly quoting the argument (yeah, "working around" it in the kernel) with `"` is preferred.

The existing use case makes just use of a subtle rule.

**Comment by jcai19** [comment]
> 1. does that solve the issue here when using `-no-integrated-as`?

Like @MaskRay mentioned, I could not figure how to make :vararg work when there are multiple directives with arguments with gnu as.

> 2. does IA recognize `:vararg` suffix?

Yes it does.

$ cat good.s
.macro foo insn1:vararg
\insn1
.endm

foo .section .foo

$llvm-mc -triple=i386 -filetype=obj good.s -o good.o
$llvm-readelf -S good.o

There are 5 section headers, starting at offset 0x60:

Section Headers:
  [Nr] Name              Type            Address  Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] .strtab           STRTAB          00000000 000044 00001c 00      0   0  1
  [ 2] .text             PROGBITS        00000000 000034 000000 00  AX  0   0  4
  [ 3] .foo              PROGBITS        00000000 000034 000000 00      0   0  1
  [ 4] .symtab           SYMTAB          00000000 000034 000010 10      1   1  4
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
  L (link order), O (extra OS processing required), G (group), T (TLS),
  C (compressed), x (unknown), o (OS specific), E (exclude),
  p (processor specific)

**Comment by MaskRay** [comment]
I have studied the integrated assembler behavior and GNU as'. Requested a documentation improvement for GNU as (https://sourceware.org/bugzilla/show_bug.cgi?id=25750#c11) and sent [[PATCH] arm64: Delete the space separator in __emit_inst](https://lore.kernel.org/linux-arm-kernel/20200413033811.75074-1-maskray@google.com/) to linux-arm-kernel

**Comment by nickdesaulniers** [comment]
accepted: https://git.kernel.org/pub/scm/linux/kernel/git/arm64/linux.git/commit/?h=for-next/fixes&id=c9a4ef66450145a356a626c833d3d7b1668b3ded

**Comment by MaskRay** [comment]
Merged into mainline: http://git.kernel.org/linus/c9a4ef66450145a356a626c833d3d7b1668b3ded
